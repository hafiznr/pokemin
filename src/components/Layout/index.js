/** @jsxRuntime classic */
/** @jsx jsx */
import React from 'react';
import { css, jsx } from '@emotion/react';
import { withRouter } from "react-router";
import { renderRoutes } from "react-router-config"
import MenuBar from '../MenuBar';

const Layout = ({route, location}) => {
  const { pathname } = location;
  const isMyPokemonPage = pathname === '/mypokemon';

  const myPokemonLinkStyles = css({
    color: isMyPokemonPage ? '#fff' : 'rgba(255,255,255,.5)',
    textDecoration: 'none',
    '&:hover, &:focus, &:active': {
      color: '#fff'
    }
  });

  const brandStyles = css({
    color: '#fff',
    textDecoration: 'none',
    '&:hover, &:focus, &:active': {
      color: '#fff',
      textDecoration: 'none'
    }
  });

  const mainStyles = css({
    maxWidth: '480px',
    boxSizing: 'border-box',
    width: '100%',
    margin: '0px auto',
    background: '#fff',
    minHeight: '100vh'
  });

  return (
    <main css={mainStyles}>
      {renderRoutes(route.routes)}
      <MenuBar pathname={pathname} />
    </main>
  )
};

export default withRouter(Layout);