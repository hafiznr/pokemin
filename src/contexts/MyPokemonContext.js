import React, { createContext, useState } from 'react';

export const MyPokemonContext = createContext();

export const MyPokemonContextProvider = ({ children, initialData }) => {
  const [myPokemonData, setMyPokemonData] = useState(JSON.parse(initialData) || []);

  const addPokemon = (data) => {
    const newPokemonData = [...myPokemonData, data];

    setMyPokemonData(newPokemonData);

    localStorage.setItem('mypokemon', JSON.stringify(newPokemonData));
  }

  const removePokemon = (nickname) => {
    const newPokemonData = myPokemonData
      .filter(pokemon => pokemon.nickname !== nickname);

    setMyPokemonData(newPokemonData);

    localStorage.setItem('mypokemon', JSON.stringify(newPokemonData));
  }

  return (
    <MyPokemonContext.Provider 
      value={{ myPokemonData, addPokemon, removePokemon }}
    >
      {children}
    </MyPokemonContext.Provider>
  )
}
