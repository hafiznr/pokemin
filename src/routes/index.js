import { lazy } from 'react';
import Layout from '../components/Layout';

const Landing = lazy(() => import('./Landing'));
const Detail = lazy(() => import('./Detail'));
const MyPokemon = lazy(() => import('./MyPokemon'));


const routes = [
  {
    component: Layout,
    routes: [
      {
        path: "/",
        exact: true,
        component: Landing
      },
      {
        path: "/pokemon/:pokemonId",
        component: Detail
      },
      {
        path: "/mypokemon",
        exact: true,
        component: MyPokemon
      }
    ]
  }
];

export default routes;