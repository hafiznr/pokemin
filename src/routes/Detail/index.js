import { useContext, useState } from 'react';
import { useParams } from 'react-router';
import { useQuery } from '@apollo/client';
import { 
  Button,
  Skeleton,
  Box
} from '@chakra-ui/react';

// queries
import { GET_POKEMON } from './queries';

// components
import SuccessModal from './components/SuccessModal';
import FailedModal from './components/FailedModal';
import Moves from './components/Moves';
import Summary from './components/Summary';

// utils
import { capitalizeFirstLetter } from '../../utils';

// context
import { MyPokemonContext } from '../../contexts/MyPokemonContext';

const Detail = () => {
  const { addPokemon } = useContext(MyPokemonContext);

  const [showModal, setShowModal] = useState('');

  const { pokemonId } = useParams();

  const { loading, error, data: pokemonData } = useQuery(GET_POKEMON, {
    variables: {
      name: pokemonId
    }
  });

  const { pokemon } = pokemonData || {};
  const {
    id,  
    moves,
    name = '', 
    sprites,
  } = pokemon || {};

  const capitalizedName = capitalizeFirstLetter(name);

  const handleClickCatch = () => {
    const probability = +process.env.REACT_APP_CATCH_PROBABILITY;
    const isSuccess = Math.random() <= probability;

    if(isSuccess) {
      setShowModal('success');
    } else {
      setShowModal('failed');
    }
  }

  const handleAddPokemon = (nickname) => {
    addPokemon({
      id,
      name,
      image: sprites.front_default,
      nickname
    });

    setShowModal('');
  }

  return (
    <Box className="detail" padding="8px 16px 78px 16px">
      {error && (
        <Box mb="8px" fontSize="xl" fontWeight="bold">
          Something went wrong. Please try again later
        </Box>
      )}

      <Box 
        textAlign="center" 
        fontSize="2xl" 
        fontWeight="bold" 
        mb="8px"
      >
        {loading ? (
          <Skeleton 
            height="27px" 
            mb="8px" 
            data-testid="name-skeleton"
          />
        ) : (
          <div data-testid="capital-name">
            {capitalizedName}
          </div>
        )}
      </Box>

      <Summary loading={loading} data={pokemon} />

      <Moves loading={loading} moves={moves} />

      <Box
        width="100%"
        textAlign="center"
        margin="16px 0"
      >
        {loading ? (
          <Skeleton 
            height="36px" 
            data-testid="catch-button-skeleton"
          />
        ) : (
          <Button 
            size="lg"
            colorScheme="twitter"
            width="100%"
            onClick={handleClickCatch}
            data-testid="catch-button"
          >
            Catch!
          </Button>
        )}
      </Box>

      <SuccessModal
        isOpen={showModal === 'success'}
        pokemon={capitalizedName}
        image={sprites?.front_default}
        onClose={() => setShowModal('')}
        onSubmit={handleAddPokemon}
      />

      <FailedModal
        isOpen={showModal === 'failed'}
        pokemon={capitalizedName}
        image={sprites?.back_default}
        onClose={() => setShowModal('')}
        onRetry={handleClickCatch}
      />
    </Box>
  );
}

export default Detail;
