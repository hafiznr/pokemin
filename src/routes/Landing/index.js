/** @jsxRuntime classic */
/** @jsx jsx */
import React, { useEffect, useState } from 'react';
import { css, jsx } from '@emotion/react';
import { useQuery, gql } from '@apollo/client';
import PokemonCard from '../../components/PokemonCard';
import { Box } from '@chakra-ui/layout';
import { useContext } from 'react';
import { MyPokemonContext } from '../../contexts/MyPokemonContext';
import TopBar from '../../components/TopBar';

const GET_POKEMONS = gql`
  query pokemons($limit: Int, $offset: Int) {
    pokemons(limit: $limit, offset: $offset) {
      count
      next
      previous
      status
      message
      results {
        id
        url
        name
        image
      }
    }
  }
`;

const Landing = () => {
  const { myPokemonData } = useContext(MyPokemonContext);
  
  const [offset, setOffset] = useState(0);
  const [pokemonList, setPokemonList] = useState([]);

  const ownedData = {};
  myPokemonData.forEach(item => {
    if (ownedData[item.id]) {
      ownedData[item.id]++;
    } else {
      ownedData[item.id] = 1;
    }
  });

  const handleSetPokemonList = (data) => {
    const { pokemons } = data || {};
    const { results } = pokemons || {};

    setPokemonList([...pokemonList, ...results]);
  }

  const { loading, error } = useQuery(GET_POKEMONS, {
    variables: {
      limit: 10,
      offset
    },
    onCompleted: handleSetPokemonList
  });

  useEffect(() => {
    const handleScroll = () => {
      if (window.innerHeight + document.documentElement.scrollTop 
        === document.documentElement.offsetHeight) {
        setOffset(offset + 10);
      }
    }

    window.addEventListener("scroll", handleScroll);

    return () => window.removeEventListener("scroll", handleScroll);
  }, [offset]);

  return (
    <React.Fragment>
      <TopBar />
      <Box padding="8px 16px 90px 16px">
        <div>
          {pokemonList.map(pokemon => {
            return (
              <PokemonCard
                id={pokemon.id}
                name={pokemon.name}
                image={pokemon.image}
                owned={ownedData[pokemon.id] || 0}
                key={pokemon.id}
              />
            )
          })}
        </div>
      </Box>
    </React.Fragment>
  );
}

export default Landing;
