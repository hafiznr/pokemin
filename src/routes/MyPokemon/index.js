import React, { useContext, useState } from 'react';
import { Box, Text } from '@chakra-ui/react';

// assets
import PokemonCard from '../../components/PokemonCard';
import PokeballIcon from '../../components/CustomIcon/PokeballIcon';

// context
import { MyPokemonContext } from '../../contexts/MyPokemonContext';

// components
import ReleaseModal from './components/ReleaseModal';

const MyPokemon = () => {
  const { myPokemonData, removePokemon } = useContext(MyPokemonContext);

  const [released, setReleased] = useState(null);

  const handleClickRelease = (pokemon) => {
    setReleased(pokemon);
  }

  const handleSubmitRelease = () => {
    removePokemon(released.nickname);

    setReleased(null);
  }

  return (
    <Box padding="8px 16px 78px 16px" data-testid="mypokemon">
      {myPokemonData.length > 0 ? (
        <div>
          {myPokemonData.map(pokemon => (
            <PokemonCard
              id={pokemon.id}
              nickname={pokemon.nickname}
              name={pokemon.name}
              image={pokemon.image}
              isMyPokemon
              onRelease={handleClickRelease}
              key={pokemon.nickname}
            />
          ))}

          <ReleaseModal
            isOpen={!!released}
            pokemon={released?.nickname}
            image={released?.image}
            onClose={() => setReleased(null)}
            onSubmit={handleSubmitRelease}
          />
        </div>
      ) : (
        <Box 
          textAlign="center" 
          height="calc(100vh - 134px)" 
          display="flex" 
          alignItems="center"
        >
          <Box width="100%">
            <PokeballIcon size="100" color="#E2E8F0"/>
            <Text fontSize="xl">
              You don't own any Pok√©mon right now...
            </Text>
          </Box>
        </Box>
      )}
    </Box>
  );
}

export default MyPokemon;
