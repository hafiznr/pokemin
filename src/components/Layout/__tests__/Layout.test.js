import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { MyPokemonContextProvider } from '../../../contexts/MyPokemonContext';
import routes from '../../../routes';
import Layout from '../';

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    pathname: "localhost:3000/"
  })
}));

test('renders Layout', () => {
  const { container } = render(
    <MyPokemonContextProvider initialData={null}>
      <BrowserRouter>
        <Layout routes={routes}/>
      </BrowserRouter>
    </MyPokemonContextProvider>
  );

  expect(container).toMatchSnapshot();
});
