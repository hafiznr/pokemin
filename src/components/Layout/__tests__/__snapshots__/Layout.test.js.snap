// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Layout 1`] = `
<div>
  <main
    class="css-1gu03f4-Layout"
  >
    <div
      class="css-1p24ppz"
    >
      <div
        class="css-1y41ry6"
        data-testid="back-button"
      >
        <svg
          fill="none"
          height="24"
          style="display: inline-block;"
          viewBox="0 0 24 24"
          width="24"
        >
          <path
            d="M16.2426 6.34317L14.8284 4.92896L7.75739 12L14.8285 19.0711L16.2427 17.6569L10.5858 12L16.2426 6.34317Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <div
        class="css-1qbkptb"
      >
        Pokémin
      </div>
    </div>
    <div
      class="css-1m966ti"
    >
      <a
        class="css-1ol0xvu"
        href="/"
      >
        <div
          class="css-13k5o8u"
        >
          <div>
            <svg
              fill="none"
              height="22"
              style="display: inline-block;"
              viewBox="0 0 24 24"
              width="22"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6 22.8787C4.34315 22.8787 3 21.5355 3 19.8787V9.87866C3 9.84477 3.00169 9.81126 3.00498 9.77823H3C3 9.20227 3.2288 8.64989 3.63607 8.24262L9.87868 2.00002C11.0502 0.828445 12.9497 0.828445 14.1213 2.00002L20.3639 8.24264C20.7712 8.6499 21 9.20227 21 9.77823H20.995C20.9983 9.81126 21 9.84477 21 9.87866V19.8787C21 21.5355 19.6569 22.8787 18 22.8787H6ZM12.7071 3.41423L19 9.70713V19.8787C19 20.4309 18.5523 20.8787 18 20.8787H15V15.8787C15 14.2218 13.6569 12.8787 12 12.8787C10.3431 12.8787 9 14.2218 9 15.8787V20.8787H6C5.44772 20.8787 5 20.4309 5 19.8787V9.7072L11.2929 3.41423C11.6834 3.02371 12.3166 3.02371 12.7071 3.41423Z"
                fill="#FEB2B2"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          Home
        </div>
      </a>
      <a
        class="css-1ol0xvu"
        href="/mypokemon"
      >
        <div
          class="css-i21swr"
        >
          <div>
            <svg
              fill="none"
              height="24"
              style="display: inline-block;"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3 12C3 7.02944 7.02944 3 12 3C16.9706 3 21 7.02944 21 12C21 16.9706 16.9706 21 12 21C7.02944 21 3 16.9706 3 12ZM5.07089 13C5.55612 16.3923 8.47353 19 12 19C15.5265 19 18.4439 16.3923 18.9291 13H14.8293C14.4174 14.1652 13.3062 15 12 15C10.6938 15 9.58251 14.1652 9.17068 13H5.07089ZM18.9291 11C18.4439 7.60771 15.5265 5 12 5C8.47353 5 5.55612 7.60771 5.07089 11H9.17068C9.58251 9.83481 10.6938 9 12 9C13.3062 9 14.4174 9.83481 14.8293 11H18.9291ZM12 13C12.5523 13 13 12.5523 13 12C13 11.4477 12.5523 11 12 11C11.4477 11 11 11.4477 11 12C11 12.5523 11.4477 13 12 13Z"
                fill="#FEB2B2"
                fill-rule="evenodd"
              />
            </svg>
          </div>
          My Pokémon (0)
        </div>
      </a>
    </div>
  </main>
</div>
`;
